python/pyarrow/src/arrow/python/arrow_to_pandas.cc

```diff
@@ -911,6 +877,122 @@ Status ConvertMap(PandasOptions options, const ChunkedArray& data,
   return Status::OK();
 }
 
+// A more helpful error message around TypeErrors that may stem from unhashable keys
+Status CheckMapAsPydictsError() {
```

Comment:
Manually tested this by remove the `with pytest.raises(TypeError):` in my non-hashable key test, and got:

```
E   TypeError: unhashable type: 'numpy.ndarray'. If keys are not hashable, then you must use `maps_as_pydicts=False`
```